# E-Commerce Analytics - Maven Fuzzy Factory

## 2.Product Analytics

This section focuses on analysing and gaining insights into the performance of individual products, evaluating the success of new product launches, and understanding cross-sell effectiveness.

In [3]:
%%sql

-- Sales growth and conversion rate growth by product type

SELECT
    YEAR(website_sessions.created_at) AS yr,
    MONTH(website_sessions.created_at) AS mo,
    COUNT(DISTINCT orders.order_id) AS total_orders,
    COUNT(DISTINCT orders.order_id)/
        COUNT(DISTINCT website_sessions.website_session_id) AS conv_rt,
    SUM(orders.price_usd)/
        COUNT(DISTINCT website_sessions.website_session_id) AS rev_per_ses,
    COUNT(DISTINCT CASE WHEN primary_product_Id = 1 THEN orders.order_id ELSE NULL END) AS product_one_orders,
    COUNT(DISTINCT CASE WHEN primary_product_Id = 2 THEN orders.order_id ELSE NULL END) AS product_two_orders
FROM website_sessions
LEFT JOIN orders
    ON website_sessions.website_session_id = orders.website_session_id
WHERE website_sessions.created_at > '2012-04-01' -- the date of the request
    AND website_sessions.created_at < '2013-04-05' -- specified in the request
GROUP BY 1,2;


 * mysql+pymysql://root:***@127.0.0.1/mavenfuzzyfactory
13 rows affected.


[(2012, 4, 99, Decimal('0.0265'), Decimal('1.325391'), 99, 0),
 (2012, 5, 108, Decimal('0.0289'), Decimal('1.445107'), 108, 0),
 (2012, 6, 140, Decimal('0.0353'), Decimal('1.765985'), 140, 0),
 (2012, 7, 169, Decimal('0.0398'), Decimal('1.988305'), 169, 0),
 (2012, 8, 228, Decimal('0.0374'), Decimal('1.869398'), 228, 0),
 (2012, 9, 287, Decimal('0.0438'), Decimal('2.191740'), 287, 0),
 (2012, 10, 371, Decimal('0.0453'), Decimal('2.266441'), 371, 0),
 (2012, 11, 618, Decimal('0.0441'), Decimal('2.204969'), 618, 0),
 (2012, 12, 506, Decimal('0.0502'), Decimal('2.511412'), 506, 0),
 (2013, 1, 391, Decimal('0.0611'), Decimal('3.127025'), 344, 47),
 (2013, 2, 497, Decimal('0.0693'), Decimal('3.692108'), 335, 162),
 (2013, 3, 385, Decimal('0.0615'), Decimal('3.176269'), 320, 65),
 (2013, 4, 96, Decimal('0.0794'), Decimal('4.085227'), 82, 14)]

Insight: Following the introduction of Product 2 (Lovebear product) in January 2013, we have observed an uptick in overall revenue per session and conversion rates. To fully grasp each product's performance, a detailed conversion funnel analysis segmented by product is required

In [4]:
%%sql

-- Analysis of conversion funnel broken down by product pages (Product 1- Mrfuzzy; Product 2 - Lovebear)

-- STEP 1: Identify pageviews with relevant sessions
-- STEP 2: Find the pageviews that occur AFTER customers saw the product page
-- STEP 3: Creat a session level conversion funnel view
-- STEP 4: Aggregate sessions to build a conversion funnel for each product page

-- STEP 1: Identify pageviews with relevant sessions
DROP TABLE IF EXISTS sessions_w_product_pages;
CREATE TEMPORARY TABLE sessions_w_product_pages
SELECT  website_session_id,
    website_pageview_id,
    pageview_url AS product_page_seen
FROM website_pageviews
WHERE created_at > '2013-01-06' -- product 2 launch date
    AND created_at < '2013-04-10' 
    AND pageview_url IN ('/the-original-mr-fuzzy', '/the-forever-love-bear');

-- STEP 2: Find the pageviews that occur AFTER customers saw the product page
SELECT DISTINCT
    website_pageviews.pageview_url
FROM
    sessions_w_product_pages
LEFT JOIN website_pageviews
    ON sessions_w_product_pages.website_session_id = website_pageviews.website_session_id
    AND website_pageviews.website_pageview_id > sessions_w_product_pages.website_pageview_id;


-- STEP 3: Creat a session level conversion funnel view
DROP TABLE IF EXISTS sessions_w_flags;    
CREATE TEMPORARY TABLE sessions_w_flags
SELECT
    website_session_id,
    CASE
        WHEN product_page_seen = '/the-original-mr-fuzzy' THEN 'mrfuzzy'
        WHEN product_page_seen = '/the-forever-love-bear' THEN 'lovebear'
        ELSE 'Check_logic'
    END AS product_seen,
    MAX(cart_page) AS cart_made_it,
    MAX(shipping_page) AS shipping_made_it,
    MAX(billing_page) AS billing_made_it,
    MAX(thankyou_page) AS thankyou_made_it
FROM (
    SELECT DISTINCT
    sessions_w_product_pages.website_session_id,
    sessions_w_product_pages.product_page_seen,
        CASE WHEN pageview_url='/cart' THEN 1 ELSE 0 END AS cart_page,
        CASE WHEN pageview_url='/shipping' THEN 1 ELSE 0 END AS shipping_page,
        CASE WHEN pageview_url='/billing-2' THEN 1 ELSE 0 END AS billing_page,
        CASE WHEN pageview_url='/thank-you-for-your-order' THEN 1 ELSE 0 END AS thankyou_page
FROM sessions_w_product_pages
LEFT JOIN website_pageviews
    ON sessions_w_product_pages.website_session_id = website_pageviews.website_session_id
    AND website_pageviews.website_pageview_id > sessions_w_product_pages.website_pageview_id
ORDER BY 1,2
) AS pageview_lvl
GROUP BY 1,
    CASE
        WHEN product_page_seen = '/the-original-mr-fuzzy' THEN 'mrfuzzy'
        WHEN product_page_seen = '/the-forever-love-bear' THEN 'lovebear'
        ELSE 'Check_logic'
    END;

    
-- STEP 4: Aggregate sessions to build a conversion funnel for each product page
SELECT 
    product_seen,
    COUNT(DISTINCT CASE WHEN cart_made_it = 1 THEN website_session_id ELSE NULL END)/
        COUNT(DISTINCT website_session_id) AS product_page_click_rt,
    COUNT(DISTINCT CASE WHEN shipping_made_it = 1 THEN website_session_id ELSE NULL END)/
        COUNT(DISTINCT CASE WHEN cart_made_it = 1 THEN website_session_id ELSE NULL END) AS cart_click_rt,
    COUNT(DISTINCT CASE WHEN billing_made_it = 1 THEN website_session_id ELSE NULL END)/
        COUNT(DISTINCT CASE WHEN shipping_made_it = 1 THEN website_session_id ELSE NULL END) AS shipping_click_rt,
    COUNT(DISTINCT CASE WHEN thankyou_made_it = 1 THEN website_session_id ELSE NULL END)/
        COUNT(DISTINCT CASE WHEN billing_made_it = 1 THEN website_session_id ELSE NULL END) AS billing_click_rt
FROM
    sessions_w_flags
GROUP BY 1;

 * mysql+pymysql://root:***@127.0.0.1/mavenfuzzyfactory
0 rows affected.
8584 rows affected.
5 rows affected.
0 rows affected.
8584 rows affected.
2 rows affected.


[('lovebear', Decimal('0.5485'), Decimal('0.6876'), Decimal('0.8093'), Decimal('0.6168')),
 ('mrfuzzy', Decimal('0.4349'), Decimal('0.6860'), Decimal('0.8205'), Decimal('0.6363'))]

Insight: Introduction of new product (Product 2- Lovebear) seems to improve the product page click-through rates. Overall, Lovebear product has a comparable clickthrough rates compared to Mrfuzzy product

In [5]:
%%sql

-- Due to the success of a new product, the company launced two more products, making a total of 4 products
-- Cross-sell Analysis

-- Monthly sales analysis by product types
-- Product 1 - Mrfuzzy
-- Product 2 - Lovebear
-- Product 3 - Birthdaybear
-- Product 4 - Minibear  


SELECT
    orders.primary_product_id,
    COUNT(DISTINCT orders.order_id) AS orders,
    COUNT(DISTINCT CASE WHEN order_items.product_id=1 THEN orders.order_id ELSE NULL END) AS x_sell_prod1,
    COUNT(DISTINCT CASE WHEN order_items.product_id=2 THEN orders.order_id ELSE NULL END) AS x_sell_prod2,
    COUNT(DISTINCT CASE WHEN order_items.product_id=3 THEN orders.order_id ELSE NULL END) AS x_sell_prod3,
    COUNT(DISTINCT CASE WHEN order_items.product_id=1 THEN orders.order_id ELSE NULL END)/COUNT(DISTINCT orders.order_id) AS x_sell_prod1_rt,
    COUNT(DISTINCT CASE WHEN order_items.product_id=2 THEN orders.order_id ELSE NULL END)/COUNT(DISTINCT orders.order_id) AS x_sell_prod2_rt,
    COUNT(DISTINCT CASE WHEN order_items.product_id=3 THEN orders.order_id ELSE NULL END)/COUNT(DISTINCT orders.order_id) AS x_sell_prod3_rt
FROM orders
    LEFT JOIN order_items
    ON order_items.order_id = orders.order_id
    AND order_items.is_primary_item = 0
GROUP BY 1;

 * mysql+pymysql://root:***@127.0.0.1/mavenfuzzyfactory
4 rows affected.


[(1, 23861, 0, 872, 1759, Decimal('0.0000'), Decimal('0.0365'), Decimal('0.0737')),
 (2, 4803, 72, 0, 136, Decimal('0.0150'), Decimal('0.0000'), Decimal('0.0283')),
 (3, 3068, 277, 112, 0, Decimal('0.0903'), Decimal('0.0365'), Decimal('0.0000')),
 (4, 581, 16, 9, 22, Decimal('0.0275'), Decimal('0.0155'), Decimal('0.0379'))]

Insight: The figures show the products that customers often purchased together with. Product 1 dominates the overall sales, with 23861 orders. There is a high cross-sell rate between product 1 and product 3. Marketing efforts should be focused on this sales bundle to capitalise on high sales potential. Also, we should explore why the cross-sell rates of other product bundles don't perform well

In [7]:
%%sql

-- Monthly sales analysis by product types
-- Product 1 - Mrfuzzy
-- Product 2 - Lovebear
-- Product 3 - Birthdaybear
-- Product 4 - Minibear  

SELECT
    YEAR(created_at) AS yr,
    MONTH(created_at) AS mon,
    SUM(price_usd) AS total_revenue,
    SUM(price_usd - cogs_usd) AS total_margin,
    SUM(CASE WHEN primary_product_id = 1 THEN price_usd ELSE NULL END) AS mrfuzzy_revenue,
    SUM(CASE WHEN primary_product_id = 1 THEN price_usd-cogs_usd ELSE NULL END) AS mrfuzzy_margin,
    SUM(CASE WHEN primary_product_id = 2 THEN price_usd ELSE NULL END) AS lovebear_revenue,
    SUM(CASE WHEN primary_product_id = 2 THEN price_usd-cogs_usd ELSE NULL END) AS lovebear_margin,
    SUM(CASE WHEN primary_product_id = 3 THEN price_usd ELSE NULL END) AS birthdaybear_revenue,
    SUM(CASE WHEN primary_product_id = 3 THEN price_usd-cogs_usd ELSE NULL END) AS birthdaybear_margin,
    SUM(CASE WHEN primary_product_id = 4 THEN price_usd ELSE NULL END) AS minibear_revenue,
    SUM(CASE WHEN primary_product_id = 4 THEN price_usd-cogs_usd ELSE NULL END) AS minibear_margin
FROM orders
WHERE created_at < '2015-02-28'
GROUP BY 1,2;

 * mysql+pymysql://root:***@127.0.0.1/mavenfuzzyfactory
36 rows affected.


[(2012, 3, Decimal('2999.40'), Decimal('1830.00'), Decimal('2999.40'), Decimal('1830.00'), None, None, None, None, None, None),
 (2012, 4, Decimal('4949.01'), Decimal('3019.50'), Decimal('4949.01'), Decimal('3019.50'), None, None, None, None, None, None),
 (2012, 5, Decimal('5398.92'), Decimal('3294.00'), Decimal('5398.92'), Decimal('3294.00'), None, None, None, None, None, None),
 (2012, 6, Decimal('6998.60'), Decimal('4270.00'), Decimal('6998.60'), Decimal('4270.00'), None, None, None, None, None, None),
 (2012, 7, Decimal('8448.31'), Decimal('5154.50'), Decimal('8448.31'), Decimal('5154.50'), None, None, None, None, None, None),
 (2012, 8, Decimal('11397.72'), Decimal('6954.00'), Decimal('11397.72'), Decimal('6954.00'), None, None, None, None, None, None),
 (2012, 9, Decimal('14347.13'), Decimal('8753.50'), Decimal('14347.13'), Decimal('8753.50'), None, None, None, None, None, None),
 (2012, 10, Decimal('18546.29'), Decimal('11315.50'), Decimal('18546.29'), Decimal('11315.50'), None, None, None, None, None, None),
 (2012, 11, Decimal('30893.82'), Decimal('18849.00'), Decimal('30893.82'), Decimal('18849.00'), None, None, None, None, None, None),
 (2012, 12, Decimal('25294.94'), Decimal('15433.00'), Decimal('25294.94'), Decimal('15433.00'), None, None, None, None, None, None),
 (2013, 1, Decimal('19966.10'), Decimal('12224.00'), Decimal('17146.57'), Decimal('10461.50'), Decimal('2819.53'), Decimal('1762.50'), None, None, None, None),
 (2013, 2, Decimal('26515.02'), Decimal('16323.00'), Decimal('16796.64'), Decimal('10248.00'), Decimal('9718.38'), Decimal('6075.00'), None, None, None, None),
 (2013, 3, Decimal('19896.15'), Decimal('12197.50'), Decimal('15996.80'), Decimal('9760.00'), Decimal('3899.35'), Decimal('2437.50'), None, None, None, None),
 (2013, 4, Decimal('28584.47'), Decimal('17524.50'), Decimal('22945.41'), Decimal('13999.50'), Decimal('5639.06'), Decimal('3525.00'), None, None, None, None),
 (2013, 5, Decimal('29364.29'), Decimal('17989.50'), Decimal('24445.11'), Decimal('14914.50'), Decimal('4919.18'), Decimal('3075.00'), None, None, None, None),
 (2013, 6, Decimal('30544.07'), Decimal('18716.50'), Decimal('25144.97'), Decimal('15341.50'), Decimal('5399.10'), Decimal('3375.00'), None, None, None, None),
 (2013, 7, Decimal('31143.96'), Decimal('19087.00'), Decimal('25444.91'), Decimal('15524.50'), Decimal('5699.05'), Decimal('3562.50'), None, None, None, None),
 (2013, 8, Decimal('31373.92'), Decimal('19230.00'), Decimal('25494.90'), Decimal('15555.00'), Decimal('5879.02'), Decimal('3675.00'), None, None, None, None),
 (2013, 9, Decimal('32723.65'), Decimal('20053.50'), Decimal('27094.59'), Decimal('16535.50'), Decimal('5629.06'), Decimal('3518.00'), None, None, None, None),
 (2013, 10, Decimal('38242.62'), Decimal('23454.00'), Decimal('31943.67'), Decimal('19516.50'), Decimal('6298.95'), Decimal('3937.50'), None, None, None, None),
 (2013, 11, Decimal('46631.02'), Decimal('28607.00'), Decimal('38192.43'), Decimal('23333.50'), Decimal('8438.59'), Decimal('5273.50'), None, None, None, None),
 (2013, 12, Decimal('58262.60'), Decimal('36190.00'), Decimal('44063.23'), Decimal('27044.50'), Decimal('8980.49'), Decimal('5615.50'), Decimal('5218.88'), Decimal('3530.00'), None, None),
 (2014, 1, Decimal('56568.89'), Decimal('35366.50'), Decimal('41049.81'), Decimal('25313.50'), Decimal('8948.49'), Decimal('5604.50'), Decimal('6570.59'), Decimal('4448.50'), None, None),
 (2014, 2, Decimal('66012.52'), Decimal('41762.00'), Decimal('36586.22'), Decimal('22822.00'), Decimal('21997.99'), Decimal('13897.50'), Decimal('7428.31'), Decimal('5042.50'), None, None),
 (2014, 3, Decimal('68189.73'), Decimal('43068.50'), Decimal('48611.65'), Decimal('30324.50'), Decimal('10718.05'), Decimal('6772.50'), Decimal('8860.03'), Decimal('5971.50'), None, None),
 (2014, 4, Decimal('78725.43'), Decimal('49713.50'), Decimal('57631.72'), Decimal('35924.00'), Decimal('10666.08'), Decimal('6719.00'), Decimal('10427.63'), Decimal('7070.50'), None, N

Insight: Sales peak during the holiday season in November and December, with Product 1 consistently leading as the top seller among the four products. Interestingly, Product 2 experiences a surge in sales every February, potentially attributable to Valentine's Day promotions and shopping trends.

In [9]:
%%sql

-- Analysing refund rates by product types

SELECT
    YEAR(order_items.created_at) AS yr,
    MONTH(order_items.created_at) AS mon,
    COUNT(DISTINCT CASE WHEN product_id = 1 THEN order_items.order_item_id ELSE NULL END) AS mrfuzzy_orders,
    COUNT(DISTINCT CASE WHEN product_id = 1 THEN order_item_refunds.order_item_id ELSE NULL END)/
        COUNT(DISTINCT CASE WHEN product_id = 1 THEN order_items.order_item_id ELSE NULL END) AS mrfuzzy_refund_rt,
    COUNT(DISTINCT CASE WHEN product_id = 2 THEN order_items.order_item_id ELSE NULL END) AS lovebear_orders,
    COUNT(DISTINCT CASE WHEN product_id = 2 THEN order_item_refunds.order_item_id ELSE NULL END)/
        COUNT(DISTINCT CASE WHEN product_id = 2 THEN order_items.order_item_id ELSE NULL END) AS lovebear_refund_rt,
    COUNT(DISTINCT CASE WHEN product_id = 3 THEN order_items.order_item_id ELSE NULL END) AS birthdaybear_orders,
    COUNT(DISTINCT CASE WHEN product_id = 3 THEN order_item_refunds.order_item_id ELSE NULL END)/
        COUNT(DISTINCT CASE WHEN product_id = 3 THEN order_items.order_item_id ELSE NULL END) AS birthdaybear_refund_rt,
    COUNT(DISTINCT CASE WHEN product_id = 4 THEN order_items.order_item_id ELSE NULL END) AS minibear_orders,
    COUNT(DISTINCT CASE WHEN product_id = 4 THEN order_item_refunds.order_item_id ELSE NULL END)/
        COUNT(DISTINCT CASE WHEN product_id = 4 THEN order_items.order_item_id ELSE NULL END) AS minibear_refund_rt
FROM order_items
    LEFT JOIN order_item_refunds
    ON order_item_refunds.order_item_id = order_items.order_item_id
WHERE order_items.created_at < '2014-10-15'
GROUP BY 1,2;

 * mysql+pymysql://root:***@127.0.0.1/mavenfuzzyfactory
32 rows affected.


[(2012, 3, 60, Decimal('0.0167'), 0, None, 0, None, 0, None),
 (2012, 4, 99, Decimal('0.0505'), 0, None, 0, None, 0, None),
 (2012, 5, 108, Decimal('0.0370'), 0, None, 0, None, 0, None),
 (2012, 6, 140, Decimal('0.0571'), 0, None, 0, None, 0, None),
 (2012, 7, 169, Decimal('0.0828'), 0, None, 0, None, 0, None),
 (2012, 8, 228, Decimal('0.0746'), 0, None, 0, None, 0, None),
 (2012, 9, 287, Decimal('0.0906'), 0, None, 0, None, 0, None),
 (2012, 10, 371, Decimal('0.0728'), 0, None, 0, None, 0, None),
 (2012, 11, 618, Decimal('0.0744'), 0, None, 0, None, 0, None),
 (2012, 12, 506, Decimal('0.0593'), 0, None, 0, None, 0, None),
 (2013, 1, 343, Decimal('0.0496'), 47, Decimal('0.0213'), 0, None, 0, None),
 (2013, 2, 336, Decimal('0.0714'), 162, Decimal('0.0123'), 0, None, 0, None),
 (2013, 3, 320, Decimal('0.0563'), 65, Decimal('0.0462'), 0, None, 0, None),
 (2013, 4, 459, Decimal('0.0414'), 94, Decimal('0.0106'), 0, None, 0, None),
 (2013, 5, 489, Decimal('0.0634'), 82, Decimal('0.0244'), 0, None, 0, None),
 (2013, 6, 503, Decimal('0.0775'), 90, Decimal('0.0556'), 0, None, 0, None),
 (2013, 7, 509, Decimal('0.0727'), 95, Decimal('0.0316'), 0, None, 0, None),
 (2013, 8, 510, Decimal('0.0549'), 98, Decimal('0.0102'), 0, None, 0, None),
 (2013, 9, 537, Decimal('0.0428'), 98, Decimal('0.0102'), 0, None, 0, None),
 (2013, 10, 603, Decimal('0.0282'), 135, Decimal('0.0148'), 0, None, 0, None),
 (2013, 11, 724, Decimal('0.0345'), 174, Decimal('0.0230'), 0, None, 0, None),
 (2013, 12, 818, Decimal('0.0232'), 183, Decimal('0.0219'), 139, Decimal('0.0719'), 0, None),
 (2014, 1, 728, Decimal('0.0426'), 183, Decimal('0.0219'), 200, Decimal('0.0650'), 0, None),
 (2014, 2, 584, Decimal('0.0394'), 351, Decimal('0.0171'), 211, Decimal('0.0664'), 202, Decimal('0.0099')),
 (2014, 3, 785, Decimal('0.0306'), 193, Decimal('0.0155'), 244, Decimal('0.0697'), 205, Decimal('0.0049')),
 (2014, 4, 917, Decimal('0.0349'), 214, Decimal('0.0187'), 267, Decimal('0.0674'), 259, Decimal('0.0154')),
 (2014, 5, 1030, Decimal('0.0291'), 246, Decimal('0.0163'), 299, Decimal('0.0569'), 298, Decimal('0.0067')),
 (2014, 6, 893, Decimal('0.0571'), 245, Decimal('0.0367'), 288, Decimal('0.0556'), 249, Decimal('0.0241')),
 (2014, 7, 961, Decimal('0.0437'), 244, Decimal('0.0369'), 276, Decimal('0.0399'), 264, Decimal('0.0152')),
 (2014, 8, 958, Decimal('0.1378'), 237, Decimal('0.0169'), 294, Decimal('0.0680'), 303, Decimal('0.0066')),
 (2014, 9, 1056, Decimal('0.1326'), 251, Decimal('0.0319'), 317, Decimal('0.0662'), 327, Decimal('0.0122')),
 (2014, 10, 513, Decimal('0.0273'), 135, Decimal('0.0074'), 165, Decimal('0.0485'), 155, Decimal('0.0323'))]

 Insight: Product 3, Birthdaybear, exhibits the highest rate of refunds. We should look into the underlying reasons for customer returns and customer dissatisfaction